### PR TITLE
Add support for ib dm (device memory)

### PR DESCRIPTION
**Pull request by roiedanino**

### CODE REVIEW

File: src/atomic_bw.c [hunk 1]
```diff
@@ -407,7 +408,7 @@ int main(int argc, char *argv[])
 	return SUCCESS;
 
 destroy_context:
-	if (destroy_ctx(&ctx,&user_param))
+	if (user_param.memory_type != MEMORY_IB_DEVICE && destroy_ctx(&ctx,&user_param))
```
Comment (roiedanino) [inline src/atomic_bw.c:411]: maybe we can just call destroy_ctx at the end

File: src/ib_device_memory.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	struct ibv_context *ib_ctx;
+	struct ibv_pd	   *pd;
+	void			   *addr;
+	void			   *malloc_address;
+};
+
+
+static int ib_memory_init(struct memory_ctx *ctx) {
+	struct ib_memory_ctx *ib_ctx = container_of(ctx, struct ib_memory_ctx, base);
+	ib_ctx->pd = ibv_alloc_pd(ib_ctx->ib_ctx);
```
Comment (roiedanino) [inline src/ib_device_memory.c:35]: maybe we should just use `ibv_open_device` to get ibv_context?

File: src/ib_device_memory.c [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	return SUCCESS;
+}
+
+static void* get_dm_mapped_address(struct ibv_dm *dm)
+{
+   	struct mlx5dv_dm dvdm = {};
+	struct mlx5dv_obj  dv = {};
+	int ret = 0;
+ 	dv.dm.in  = dm;
+    dv.dm.out = &dvdm;
```
Comment (roiedanino) [inline src/ib_device_memory.c:54]: align

File: src/ib_device_memory.c [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	ctx->base.destroy = ib_memory_destroy;
+	ctx->base.allocate_buffer = ib_memory_allocate_buffer;
+	ctx->base.free_buffer = ib_memory_free_buffer;
+	ctx->base.copy_host_to_buffer = memcpy;
+	ctx->base.copy_buffer_to_host = memcpy;
+	ctx->base.copy_buffer_to_buffer = memcpy;
+	ctx->ib_ctx = params->ib_ctx;
+
+	return &ctx->base;
+}
```
Comment (roiedanino) [inline src/ib_device_memory.c:120]: blank line

File: src/ib_device_memory.h [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	return false;
+}
+
+inline struct memory_ctx *ib_memory_create(struct perftest_parameters *params) {
+	return NULL;
+}
+
+#endif
+
+#endif /* IB_DEVICE_MEMORY_H */
```
Comment (roiedanino) [inline src/ib_device_memory.h:32]: blank line

File: src/perftest_parameters.c [hunk 6]
```diff
@@ -2325,6 +2327,7 @@ int parser(struct perftest_parameters *user_param,char *argv[], int argc)
 			{ .name = "use_cuda_bus_id",	.has_arg = 1, .flag = &use_cuda_bus_id_flag, .val = 1},
 			{ .name = "use_cuda_dmabuf",	.has_arg = 0, .flag = &use_cuda_dmabuf_flag, .val = 1},
 			{ .name = "use_rocm",		.has_arg = 1, .flag = &use_rocm_flag, .val = 1},
+			{ .name = "use_ib_dm",	    .has_arg = 0, .flag = &use_ib_dm_flag, .val = 1},
```
Comment (roiedanino) [inline src/perftest_parameters.c:2330]: align

File: src/perftest_parameters.c [hunk 7]
```diff
@@ -2785,6 +2788,12 @@ int parser(struct perftest_parameters *user_param,char *argv[], int argc)
 					user_param->memory_create = rocm_memory_create;
 					use_rocm_flag = 0;
 				}
+				if (use_ib_dm_flag) {
+					user_param->memory_type 	= MEMORY_IB_DEVICE;
+					user_param->memory_create 	= ib_memory_create;
+					user_param->use_ib_dm		= ON;
```
Comment (roiedanino) [inline src/perftest_parameters.c:2794]: too many spaces

File: src/perftest_parameters.h [hunk 8]
```diff
@@ -631,6 +632,8 @@ struct perftest_parameters {
 	char				*source_ip;
 	int 				has_source_ip;
 	int 			ah_allocated;
+	struct ibv_context   *ib_ctx;
```
Comment (roiedanino) [inline src/perftest_parameters.h:635]: try to align

File: src/ib_device_memory.c [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	ret = mlx5dv_init_obj(&dv, MLX5DV_OBJ_DM);
+	if (ret != 0) {
+		return NULL;
+    }
+    return dvdm.buf;
+}
+
+static int ib_memory_allocate_buffer(struct memory_ctx *ctx, int alignment, uint64_t size, int *dmabuf_fd,
+									 uint64_t *dmabuf_offset,  void **addr, bool *can_init) {
+#if HAVE_IBV_DM
```
Comment (roiedanino) [inline src/ib_device_memory.c:64]: probably not necessary

File: configure.ac [hunk 10]
```diff
@@ -277,6 +277,21 @@ AM_CONDITIONAL([CUDA_DMA_BUF_PARAMETERS_SUPPORT],[test "x$CUDA_DMA_BUF_PARAMETER
...
+            [AS_HELP_STRING([--with-dm], [Compile with Infiniband Device Memory support])],
+            [],
+            [with_dm=yes])
+
+AS_IF([test "x$with_dm" != xno], [
+    AC_CHECK_DECLS([ibv_alloc_dm],
+        [AC_DEFINE([HAVE_IBV_DM], 1, [Device Memory support])],
+        [], [[#include <infiniband/verbs.h>]])])
+
+#TODO: Change condition to check if HAVE_IBV_DM defined
```
Comment (roiedanino) [inline configure.ac:291]: Fix

**Comment by HassanKhadour** [comment]
Hi @roiedanino,
Thanks for your contribution.
Please add more info to the commit message

**Pull request closed without merging**